README.txt

Helpx

README

CONTENTS OF THIS FILE
---------------------
01| HOW TO USE THIS DOCUMENT
02| INTRO
03| GETTING STARTED
04| DEPENDENCIES
05| BUILDING
06| INSTALLATION
07| CONFIGURATION
08| MANUAL
09| FAQ
10| PLATFORM NOTES
11| TROUBLESHOOTING
12| KNOWN ISSUES
13| BUG REPORTING
14| FEEDBACK
15| TESTING
16| CONTRIBUTING
17| UPDATING
18| RECENT CHANGES
19| LICENSE
20| LEGAL
21| CREDITS


---01| HOW TO USE THIS DOCUMENT

(Note that this README is designed to be filled out over the course of many
versions. Some sections are currently blank (e.g. Recent Changes) because it is
early in the project.)

You might prefer to use the website (http://www.robertbrogan.com/helpx) for
better and more up-to-date info instead. Otherwise, prefer to use HTML versions
as they are hyperlinked.

What is the software about?
See 01|INTRO.

Is it OK for me to use? OK for me to modify? OK to make copies?
See 19|LICENSE for info about the software license. License itself is in
LICENSE.txt.
See 20|LEGAL for any additional info.

How do I get it working?
See 03|GETTING STARTED (after you have it installed and configured) to see how
to use.
See 04|DEPENDENCIES for anything external that is required or optional.
See 05|BUILDING for how to compile from source (if even necessary).
See 06|INSTALLATION for how to install it on your system (and how to uninstall).
See 07|CONFIGURATION for how you can customize it for your own use.

I cannot make it work, what now?
See 11|TROUBLESHOOTING for dealing with problems with the software.
See 10|PLATFORM NOTES for ensuring it works with your platform/OS.
See 08|MANUAL to make sure you are using it correctly.
See 09|FAQ to see if your question has been answered.
See 12|KNOWN ISSUES to see if your problem is already known about (and any
workarounds / advice).
See 13|BUG REPORTING if you want to make a report and get follow-up.

What is in the other sections?
14|FEEDBACK - Info about things like feature requests.
15|TESTING - How you can test changes you make to the code.
16|CONTRIBUTING - How you improve the product for everyone.
17|UPDATING - How to get the latest changes.
18|RECENT CHANGES - What the latest changest are.
21|CREDITS - Who or what helped with making Helpx.

---02| INTRO

One of the nicest features about Tcl is its command documentation. Likewise, the
man pages for Linux can be very helpful and convenient. Would you like to add
such features to your Tcl programs?

Helpx will let you generate HTML pages that look like the core Tcl
documentation. Helpx also has a module that allows your users to pull up command
documentation right from the Tcl shell.

---03| GETTING STARTED

If you would like to confirm the library is installed and working then try the
following from a Tcl shell.

% package require helpx
0.2
% helpx helpx
SIGNATURE
    helpx {args}

DESCRIPTION
    Print out help info for given command.

RETURNS
    None.

ARGUMENTS
    args
         Command name or option flag (see More Info for list). Last argument will
be taken as the name of the command.

MORE INFO
    If you use no flags or -all then helpx will show all info. If you want
specific info only, then you can use one or more of the following flags.

    -arg : Arguments.

    -more : More Info section.

    -desc : Short (<= 60 words) description.

    -ex : Examples.

    -ret : Return Value.

    -sig : Signature.

    -see : See Also



SEE ALSO


EXAMPLE(S)

% helpx -sig String2File
SIGNATURE
String2File {StringValue OutFilePath}

You should be all set!

---04| DEPENDENCIES

Helpx uses the following packages, which you likely will already have got in
your Tcl distribution:

    sqlite3
    Tclx
    textutil::string
    textutil::split

If you do not have these, then check the documentation that came with your
distro for information on how to get them.

---05| BUILDING

Helpx is provided as a simple Tcl package and does not need to be built.

---06| INSTALLATION

If you are reading this you most likely have already successfully installed.

To install, extract the contents of the archive file (.zip or .tar.gz). That
will result in the following files:


README.txt
     Plenty of info on how to get going.
LICENSE.txt
     Terms of use and whatnot.
WARNING.txt
     Special notices to prevent surprises.
/src
     Source files
     pkgIndex.tcl
          Used by the TCL package mechanism.
     helpx.tcl
          Main script file.
/doc
     Any other documents (like the manual).
/test
     Test suite

You will probably want to add the following at the bottom of your init.tcl so
that Tcl can find the helpx package:

lappend auto_path YOUR_DIR_PATH

See 04|DEPENDENCIES for info about what Helpx may need.

See 07|CONFIGURATION for more info about init.tcl.

What to do after you install? You can check out the 03|GETTING STARTED section
of the README. We recommend browsing the manual (or even better, use the
helpx.chm file included in your download). Each command reference has at least
one example. You can type that directly into your terminal and try out the
command.

Prefer use to git and clone the repo? It is posted up for you at the following
locations: https://github.com/rbrogan/tcl-helpx

How to uninstall? To uninstall you can simply delete the directory. You can also
remove the lappend auto_path YOUR_DIR_PATH line from your Tcl init.

---07| CONFIGURATION

You will probably want to add the following at the bottom of your init.tcl so
that Tcl can find the helpx package:

lappend auto_path YOUR_DIR_PATH

How to find init.tcl? Here are some locations found for Windows and Linux. Most
likely if you use the number of your version then you can find your init.tcl in
a similar spot.

C:\Tcl\lib\tcl8.5\init.tcl
C:\Tcl\lib\tcl8.6\init.tcl
/usr/share/tcltk/tcl8.4/init.tcl
/usr/share/tcltk/tcl8.5/init.tcl

Alternatively, do a search for init.tcl starting from the root directory of your
installation.

Be sure to check this space when SQL-related commands are added.

---08| MANUAL

We have a few options. You can see the online manual at
http://www.robertbrogan.com/helpx/doc/manual-home.html (which will be the most
up-to-date).

Alternatively, you can use the offline version at doc/manual-home.html or
(preferrably) use the compiled HTML Help version at doc/helpx.chm.

---09| FAQ

No questions yet. We will put them here as we get them.

Please send questions you have to :

helpx.questions@robertbrogan.com or you can visit
http://www.robertbrogan.com/helpx/feedback.html and use the form there.

Also note, you may possibly find the answer to your question in 08|MANUAL,
10|PLATFORM NOTES, 11|TROUBLESHOOTING, or 12|KNOWN ISSUES.

---10| PLATFORM NOTES

If you have Tcl working for your platform, then the library should work without
problem.

None of the commands themselves include platform-specifc interactions (i.e. make
system calls) beyond occasionally calling another Tcl command that is
platform-specific (e.g. in an upcoming release the command Reg2Dict will call
the Tcl command registry, which is Windows-specific).

---11| TROUBLESHOOTING

No tips at this time.

Also note, you may possibly find help in 07|MANUAL, 09|PLATFORM NOTES, 08|FAQ,
or 11|KNOWN ISSUES.

---12| KNOWN ISSUES

None at this time.

None at this time. We will post them here as they arise.

---13| BUG REPORTING

Visit http://www.robertbrogan.com/helpx/feedback.html.

Alternatively, send an email to one of:

helpx.questions@robertbrogan.com
helpx.comments@robertbrogan.com
helpx.bugreport@robertbrogan.com
helpx.wishlist@robertbrogan.com
helpx.other@robertbrogan.com

and we will try to get back to you ASAP.

---14| FEEDBACK

You can visit our feedback page at --
http://www.robertbrogan.com/helpx/feedback.html.

Alternatively, send an email to one of:

helpx.questions@robertbrogan.com
helpx.comments@robertbrogan.com
helpx.bugreport@robertbrogan.com
helpx.wishlist@robertbrogan.com
helpx.other@robertbrogan.com

and we will try to get back to you ASAP.

---15| TESTING

You can find tests in the /test directory.

You will find a README file there as well, further details on things like how to
run the tests yourself.

---16| CONTRIBUTING

Nothing formal has been set up for governing this project, yet.

---17| UPDATING

The latest version can be found at
http://www.robertbrogan.com/helpx/download.html.

Want to receive notifications about changes to Helpx? You can subscribe to the
announcements mailing list by sending an email to
helpx-announce-subscribe@robertbrogan.com. (No need for anything in subject or
message body.) You will get an email every time we have a new release.

---18| RECENT CHANGES

Pre-release to support Gen.

---19| LICENSE

The license is the same as the license for Tcl, for all practical purposes. See
LICENSE.txt or license.html for the actual license.

---20| LEGAL

No legal notice at this time (i.e. no use of crypto). See LICENSE.txt or
license.html for the license.

---21| CREDITS

Information posted at wiki.tcl.tk has been helpful throughout work on Tcl
projects.



news.txt
Helpx

NEWS

2015.05.11
v0.2.0 Release

For this release, we make available the API for you to use in making your own
Helpx databases and bundle them with your packages.

HelpxNS::ArgumentInfoFor
     Get dict with argument names and descriptions for command.
     
HelpxNS::CreateCommand
     Create a new command entry.
     
HelpxNS::CreateHelpxDatabase
     Create a database file and populate it with helpx tables.
     
HelpxNS::DeleteCommand
     Delete a command entry and all associated info.
     
HelpxNS::EnterArgumentInfo
     Insert or update argument info for a command.
     
HelpxNS::EnterExample
     Create / update an example showing command in use.
     
HelpxNS::EnterMoreInfo
     Enter what is to appear in the More Info section.
     
HelpxNS::EnterReturnValue
     Enters information about the return value for the given proc.
     
HelpxNS::EnterSeeAlso
     Enter other commands to make links to.
     
HelpxNS::EnterShortDescription
     Enters a short description (<= 60 characters) for the given proc.
     
HelpxNS::EnterSignature
     Enters the signature for the given proc.
     
HelpxNS::EraseArgumentInfo
     Delete argument and adjust positions.
     
HelpxNS::EraseExample
     Erase example of command, at given position.
     
HelpxNS::EraseSeeAlso
     Erase see also link, at given position.
     
HelpxNS::ExamplesFor
     Get an ordered list of examples showing command in use.
     
HelpxNS::MoreInfoFor
     Get additional information for given command.
     
HelpxNS::RenameCommand
     Rename the command to a new name.
     
HelpxNS::ReturnValueFor
     Gets a description of the return value for the given proc.
     
HelpxNS::SeeAlsoFor
     Get a list of commands related to the given command.
     
HelpxNS::ShortDescriptionFor
     Returns the shorter description for the given proc.
     
HelpxNS::SignatureFor
     Returns the signature for the given proc.
     

2015.04.06
v0.1.0 Release

Pre-release to support Gen.


src/helpx.tcl
package provide helpx 0.2
package require sqlite3
package require Tclx
package require textutil::string
package require textutil::adjust
package require struct::list
package require  gen
if {[string equal $::tcl_platform(platform) "windows"]} {
     package require registry
}

array set ErrorCode {
     VARIABLE_NOT_FOUND -1
     INPUT_NON_NUMERIC -2
     VARIABLE_NAME_EMPTY -3
     INDEX_OUT_OF_RANGE -4
     INPUT_NOT_WELL_FORMED -5
     CANNOT_FACTOR_INPUT_LIST -6
     INPUT_INVALID -7
     INPUT_OUT_OF_RANGE -8
     INPUT_NON_POSITIVE -9
     SEARCH_STRING_EMPTY -10
     VARIABLE_CONTENTS_INVALID -11
     VARIABLE_CONTENTS_EMPTY -12
     DATABASE_VARIABLE_NOT_FOUND -13
     TABLE_NOT_FOUND -14
     ARGUMENTS_INCOHERENT -15
     REGISTRY_ELEMENT_NOT_FOUND -16
     PROC_NOT_FOUND -17
     ALREADY_EXISTS -18
}

array set ErrorMessage {
     VARIABLE_NOT_FOUND {Could not find variable %s in caller.}
     INPUT_NON_NUMERIC {Got input value %s. Expected numeric value.}
     VARIABLE_NAME_EMPTY {Variable name is missing. Got empty string.}
     INDEX_OUT_OF_RANGE {List index %s is invalid.}
     INPUT_NOT_WELL_FORMED {Got input value %s. Expected input of the form %s.}
     CANNOT_FACTOR_INPUT_LIST {Input value %s does not divide list evenly.}
     INPUT_INVALID {Input value %s is invalid.}
     INPUT_OUT_OF_RANGE {Input value %s is out-of-range.}
     INPUT_NON_POSITIVE {Expected positive input value, got input value %s instead.}
     SEARCH_STRING_EMPTY {Got empty string for search value.}
     VARIABLE_CONTENTS_INVALID {Variable %s has invalid value %s.}
     VARIABLE_CONTENTS_EMPTY {Variable %s has empty value.}
     DATABASE_VARIABLE_NOT_FOUND {No variable called %s was found in the database globals table.}
     TABLE_NOT_FOUND {Table %s not found.}
     ARGUMENTS_INCOHERENT {Arguments %s and %s have incoherent values %s and %s.}
     REGISTRY_ELEMENT_NOT_FOUND {Registry key/value %s not found.}
     PROC_NOT_FOUND {Could not find proc %s.}
     ALREADY_EXISTS {Value %s in variable %s already exists.}
}


namespace eval HelpxNS {
     variable DatabaseConnectionList {}
}
proc HelpxNS::ArgumentInfoFor Command {

     set CommandId [CommandIdFor $Command]
     set ::sql "SELECT argname, argdesc FROM command_args WHERE commandid = $CommandId ORDER BY position ASC"
     set Dict [dict create {*}[QQ $::sql]]
     return $Dict
}

proc HelpxNS::CommandIdFor Command {

     if {[string is integer $Command]} {
          return $Command
     } else {
          set ::sql "SELECT id FROM commands WHERE name = '$Command'"
          set CommandId [Q1 $::sql]
          if {[IsEmpty $CommandId]} {
               error "Command $Command not found."
          } else {
               return $CommandId
          }
     }
}

proc HelpxNS::CreateCommand Name {

     if {[IsEmpty $Name]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Name] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     set ::sql "SELECT id FROM commands WHERE name = '$Name'"
     set CommandId [Q1 $::sql]
     if {[NotEmpty $CommandId]} {
          error [format $::ErrorMessage(ALREADY_EXISTS) $Name Name] $::errorInfo $::ErrorCode(ALREADY_EXISTS)
     }

     set ::sql "INSERT INTO commands (name) VALUES ('$Name')"
     QQ $::sql
     return [LastId commands]
}

proc HelpxNS::CreateHelpxDatabase FilePath {

     if {[IsEmpty $FilePath]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) FilePath] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }

     sqlite3 helpxdb $FilePath
     set TableDict  [dict create   commands {  {id integer primary key}  {name text}  }  commandinfo {  {id integer primary key}  {commandid integer}  {infotype text}  {infodata text}  }  command_args {  {id integer primary key}  {commandid integer}  {argname text}  {argdesc text} {position integer} }  command_examples {  {id integer primary key}  {commandid integer}  {data text} {position integer}  }  command_seealso {{id integer primary key} {commandid integer} {data text} {position integer}}]
     dict for {TableName TableDef} $TableDict {
          set sql "SELECT count(*) FROM sqlite_master WHERE tbl_name = '$TableName'"
          set TableExists [lindex [helpxdb eval $sql] 0]
          if {$TableExists} {
               helpxdb eval "DROP TABLE $TableName"
          }
          helpxdb eval "CREATE TABLE $TableName ([join $TableDef ","])"
     }    
}

proc HelpxNS::DeleteCommand Command {

     set CommandId [CommandIdFor $Command]
     set ::sql "DELETE FROM commands WHERE id = $CommandId"
     QQ $::sql

     # Delete from command_args table
     set ::sql "DELETE FROM command_args WHERE commandid = $CommandId"
     QQ $::sql

     # Delete from command_examples table
     set ::sql "DELETE FROM command_examples WHERE commandid = $CommandId"
     QQ $::sql

     # Delete from commandinfo table
     set ::sql "DELETE FROM commandinfo WHERE commandid = $CommandId"
     QQ $::sql
}

proc HelpxNS::EnterArgumentInfo {Command ArgName ArgDesc {Position -1}} {

     if {[IsEmpty $ArgName]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) ArgName] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     set CommandId [CommandIdFor $Command]
     set ::sql "SELECT position FROM command_args WHERE commandid = $CommandId AND argname = '$ArgName'"
     set CurrentPosition [Q1 $::sql]
     if {[NotEmpty $CurrentPosition]} {
          if {($Position == -1)} {
               set Position $CurrentPosition
          } elseif {$Position != $CurrentPosition} {
               error [format $::ErrorMessage(ALREADY_EXISTS) $ArgName ArgName] $::errorInfo $::ErrorCode(ALREADY_EXISTS)
          }
     } elseif {$Position == -1} {
          # If it is command name then get the ID     
          # No position was given so we will figure out what should be the last and use that.
          set ::sql "SELECT max(position) FROM command_args WHERE commandid = $CommandId"
          set Position [Q1 $::sql]
          if {[IsEmpty $Position]} {
               set Position -1
          }
          incr Position
     }
     if {[IsEmpty $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     if {![IsNumeric $Position] || ![IsNonNegative $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_INVALID) Position $Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_INVALID)
     }

     return [RunSqlEnter command_args [dict create commandid $CommandId  position $Position] [dict create argname "'$ArgName'" argdesc "'$ArgDesc'"]]
}

proc HelpxNS::EnterExample {Command ExampleData {Position -1}} {

     set CommandId [CommandIdFor $Command]
     # If it is command name then get the ID
     if {$Position == -1} {
          # No position was given so we will figure out what should be the last and use that.
          set ::sql "SELECT max(position) FROM command_examples WHERE commandid = $CommandId"
          set Position [Q1 $::sql]
          if {[IsEmpty $Position]} {
               set Position -1
          }
          incr Position
     } elseif {[IsEmpty $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     } elseif {![IsNumeric $Position] || ![IsNonNegative $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_INVALID) Position $Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_INVALID)
     }

     return [RunSqlEnter command_examples [dict create commandid $CommandId position $Position] [dict create data "'$ExampleData'"]]
}

proc HelpxNS::EnterInfoData {Command InfoType InfoData} {

     return [RunSqlEnter commandinfo [dict create commandid [CommandIdFor $Command] infotype "'$InfoType'"] [dict create infodata "'$InfoData'"]]

}

proc HelpxNS::EnterMoreInfo {Command MoreInfo} {

     return [EnterInfoData [CommandIdFor $Command] moreinfo $MoreInfo]
}

proc HelpxNS::EnterReturnValue {Command ReturnValueDescription} {

     return [EnterInfoData [CommandIdFor $Command] returnvalue $ReturnValueDescription]
}

proc HelpxNS::EnterSeeAlso {Command SeeAlsoData {Position -1}} {

     set CommandId [CommandIdFor $Command]
     if {$Position == -1} {
          # No position was given, so we will figure out what should be last and use that.
          set ::sql "SELECT max(position) FROM command_seealso WHERE commandid = $CommandId"
          set Position [Q1 $::sql]
          if {[IsEmpty $Position]} {
               set Position -1
          }
          incr Position
     } elseif {[IsEmpty $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     } elseif {![IsNumeric $Position] || ![IsNonNegative $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_INVALID) Position $Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_INVALID)
     }

     return [RunSqlEnter command_seealso [dict create commandid $CommandId position $Position] [dict create data "'$SeeAlsoData'"]]

}

proc HelpxNS::EnterShortDescription {Command ShortDescription} {


     return [EnterInfoData [CommandIdFor $Command] sdesc $ShortDescription]
}

proc HelpxNS::EnterSignature {Command Signature} {

     return [EnterInfoData [CommandIdFor $Command] signature $Signature]
}

proc HelpxNS::EraseArgumentInfo {Command Position} {

     if {[IsEmpty $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     if {![IsNumeric $Position] || ![IsNonNegative $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_INVALID) Position $Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_INVALID)
     }
     set CommandId [CommandIdFor $Command]

     # Determine if position is out-of-range
     set ::sql "SELECT max(position) FROM command_args WHERE commandid = $CommandId"
     set MaxPosition [Q1 $::sql]
     if {$Position > $MaxPosition} {
          error [format $::ErrorMessage(INPUT_OUT_OF_RANGE) $Position] $::errorInfo $::ErrorCode(INPUT_OUT_OF_RANGE)          
     }

     set ::sql "DELETE FROM command_args WHERE commandid = $CommandId AND position = $Position"
     QQ $::sql

     # Update positions so there are no gaps
     set LastPosition -1
     foreach {Id Position} [QQ "SELECT id, position FROM command_args WHERE commandid = $CommandId ORDER BY position ASC"] {
          set ExpectedPosition [expr $LastPosition + 1]
          if {$Position != $ExpectedPosition} {
               set ::sql "UPDATE command_args SET position = $ExpectedPosition WHERE id = $Id"
               QQ $::sql
          }
          set LastPosition $ExpectedPosition
     }
}

proc HelpxNS::EraseExample {Command Position} {

     if {[IsEmpty $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     if {![IsNumeric $Position] || ![IsNonNegative $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_INVALID) Position $Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_INVALID)
     }

     set CommandId [CommandIdFor $Command]

     # Determine if position is out-of-range
     set ::sql "SELECT max(position) FROM command_examples WHERE commandid = $CommandId"
     set MaxPosition [Q1 $::sql]
     if {$Position > $MaxPosition} {
          error [format $::ErrorMessage(INPUT_OUT_OF_RANGE) $Position] $::errorInfo $::ErrorCode(INPUT_OUT_OF_RANGE)          
     }

     set ::sql "DELETE FROM command_examples WHERE commandid = $CommandId AND position = $Position"
     QQ $::sql

     # Update positions so there are no gaps
     set LastPosition -1
     foreach {Id Position} [QQ "SELECT id, position FROM command_examples WHERE commandid = $CommandId ORDER BY position ASC"] {
          set ExpectedPosition [expr $LastPosition + 1]
          if {$Position != $ExpectedPosition} {
               set ::sql "UPDATE command_examples SET position = $ExpectedPosition WHERE id = $Id"
               QQ $::sql
          }
          set LastPosition $ExpectedPosition
     }
}

proc HelpxNS::EraseSeeAlso {Command Position} {

     if {[IsEmpty $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     if {![IsNumeric $Position] || ![IsNonNegative $Position]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_INVALID) Position $Position] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_INVALID)
     }

     set CommandId [CommandIdFor $Command]

     # Determine if position is out-of-range
     set ::sql "SELECT max(position) FROM command_seealso WHERE commandid = $CommandId"
     set MaxPosition [Q1 $::sql]
     if {$Position > $MaxPosition} {
          error [format $::ErrorMessage(INPUT_OUT_OF_RANGE) $Position] $::errorInfo $::ErrorCode(INPUT_OUT_OF_RANGE)          
     }

     set ::sql "DELETE FROM command_seealso WHERE commandid = $CommandId AND position = $Position"
     QQ $::sql

     # Update positions so there are no gaps
     set LastPosition -1
     foreach {Id Position} [QQ "SELECT id, position FROM command_seealso WHERE commandid = $CommandId ORDER BY position ASC"] {
          set ExpectedPosition [expr $LastPosition + 1]
          if {$Position != $ExpectedPosition} {
               set ::sql "UPDATE command_examples SET position = $ExpectedPosition WHERE id = $Id"
               QQ $::sql
          }
          set LastPosition $ExpectedPosition
     }
}

proc HelpxNS::ExamplesFor Command {

     set ::sql "SELECT data FROM command_examples WHERE commandid = [CommandIdFor $Command] ORDER BY position ASC"
     return [QQ $::sql]
}

proc HelpxNS::InfoDataFor {Command InfoType} {

     set ::sql "SELECT infodata FROM commandinfo WHERE infotype = '$InfoType' AND commandid = [CommandIdFor $Command]"
     return [QQ $::sql]
}

proc HelpxNS::MoreInfoFor Command {

     return [struct::list flatten [InfoDataFor [CommandIdFor $Command] moreinfo]]
}

proc HelpxNS::MultiDatabaseQuery Query {

     variable DatabaseConnectionList
     
     foreach Connection $DatabaseConnectionList {
          set Results [$Connection eval $Query]
          if {[$Results ne ""]} {
               return $Results
          }
     }
     
     return ""
}

proc HelpxNS::RegisterDatabase ConnectionName {

     variable DatabaseConnectionList

     if {[IsEmpty $ConnectionName]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) ConnectionName] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }
     
     if {[lsearch $DatabaseConnectionList $ConnectionName] == -1} {
          lappend DatabaseConnectionList $ConnectionName
     }
}

proc HelpxNS::RenameCommand {Command NewName} {

     if {[IsEmpty $NewName]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) NewName] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }

     if {[SqlRecordExists commands [dict create name "'$NewName'"]]} {
          error [format $::ErrorMessage(ALREADY_EXISTS) $NewName NewName] $::errorInfo $::ErrorCode(ALREADY_EXISTS)
     }

     set ::sql "UPDATE commands SET name = '$NewName' WHERE id = [CommandIdFor $Command]"
     QQ $::sql
}

proc HelpxNS::ReturnValueFor Command {

     return [struct::list flatten [InfoDataFor [CommandIdFor $Command] returnvalue]]
}

proc HelpxNS::SeeAlsoFor Command {

     set ::sql "SELECT data FROM command_seealso WHERE commandid = [CommandIdFor $Command] ORDER BY position ASC"
     set SeeAlsoList [QQ $::sql]
     foreach Element $SeeAlsoList {
          set ::sql "SELECT id FROM commands WHERE name = '$Command'"
          set CommandId [Q1 $::sql]
          if {[IsEmpty $CommandId]} {
               continue
          }
     }

     return $SeeAlsoList
}

proc HelpxNS::ShortDescriptionFor Command {

     return [struct::list flatten [InfoDataFor [CommandIdFor $Command] sdesc]]
}

proc HelpxNS::SignatureFor Command {

     return [struct::list flatten [InfoDataFor [CommandIdFor $Command] signature]]
}

proc HelpxNS::UnregisterDatabase ConnectionName {

     variable DatabaseConnectionList

     if {[IsEmpty $ConnectionName]} {
          error [format $::ErrorMessage(VARIABLE_CONTENTS_EMPTY) ConnectionName] $::errorInfo $::ErrorCode(VARIABLE_CONTENTS_EMPTY)
     }


     while {[set Index [lsearch $DatabaseConnectionList $ConnectionName]] != -1} {
          set DatabaseConnectionList [lreplace $DatabaseConnectionList $Index $Index]
     }
}

proc helpx args {

     # If second argument is not used or if -all flag is used,
     # then show all the information,
     # otherwise, just show the piece corresponding to the flag.
     if {([llength $args] == 1) || ([lindex $args 0] eq "-all")} {
          set Options [list -arg -more -desc -ex -ret -sig -see]
     } else {
          set Options [lrange $args 0 end-1]
     }
     set CommandName [lindex $args end]
     
    
     # In future, figure out how to deal with multiple commands with same name.
     set Results ""
     set ::sql "SELECT id FROM commands WHERE name = '$CommandName'"
     foreach ConnectionName $HelpxNS::DatabaseConnectionList {
          set Results [$ConnectionName eval $::sql]
          if {$Results ne ""} {
               set TargetConnection $ConnectionName
               set CommandId [lindex $Results 0]
               break
          }
     }
     if {$Results eq ""} {
          puts "$CommandName not found."
          return
     }
     
     set InfoTypes {example more desc signature returnvalue seealso}

     # Based on the options selected, get the data from the database.
     set InfoDict [dict create]
     foreach Option $Options {
          switch $Option {
               -arg {
                    set InfoType args
                    set ::sql "SELECT argname, argdesc FROM command_args WHERE commandid = $CommandId ORDER BY position ASC"
                    dict set InfoDict arguments [$TargetConnection eval $::sql]                    
               }
               -more {
                    set InfoType more
                    set ::sql "SELECT infodata FROM commandinfo WHERE infotype = 'moreinfo' AND commandid = $CommandId"
                    dict set InfoDict more [lindex [$ConnectionName eval $::sql] 0]

               }
               -desc {
                    set InfoType desc
                    set ::sql "SELECT infodata FROM commandinfo WHERE infotype = 'sdesc' AND commandid = $CommandId"
                    dict set InfoDict desc [lindex [$ConnectionName eval $::sql] 0]
               }          
               -ex {
                    set InfoType example
                    set ::sql "SELECT data FROM command_examples WHERE commandid = $CommandId ORDER BY position ASC"
                    dict set InfoDict example [$ConnectionName eval $::sql]
               }
               -ret {
                    set InfoType returnvalue
                    set ::sql "SELECT infodata FROM commandinfo WHERE infotype = 'returnvalue' AND commandid = $CommandId"
                    dict set InfoDict returnvalue [lindex [$ConnectionName eval $::sql] 0]
               }
               -sig {              
                    set InfoType signature
                    set ::sql "SELECT infodata FROM commandinfo WHERE infotype = 'signature' AND commandid = $CommandId"
                    dict set InfoDict signature [lindex [$ConnectionName eval $::sql] 0]
               }
               -see {
                    set InfoType seealso
                    set ::sql "SELECT infodata FROM commandinfo WHERE infotype = 'seealso' AND commandid = $CommandId"
                    dict lappend InfoDict seealso [$ConnectionName eval $::sql]
               }
               default {
                    error "Unknown option $Option"
               }
          }     
     }
     # Set the order in which each element will be displayed.
     set DisplayOrder {signature desc returnvalue arguments more seealso example}
     
     # Print each element, in turn.
     foreach key $DisplayOrder {
          if {![dict exists $InfoDict $key]} {
               continue
          }
          set value [dict get $InfoDict $key]
          set value [regsub -all {(<.+?>)} $value ""]
          set value [regsub -all {(</.+?>)} $value ""]
          set value [regsub -all {\\&lt;} $value <]
          set value [regsub -all {\\&gt;} $value >]
          
          switch $key {
               desc {
                    puts "DESCRIPTION"
                    puts "    [::textutil::adjust::adjust $value]"
               }
               more {
                    puts "MORE INFO"
                    set Paragraphs [split $value "\n"]
                    foreach Paragraph $Paragraphs {
                         puts "    [::textutil::adjust::adjust $Paragraph]"
                    }
               }
               signature {
                    puts "SIGNATURE"
                    puts "    [::textutil::adjust::adjust $value]"
               }
               returnvalue {
                    puts "RETURNS"
                    foreach Element [split $value "|"] {
                         puts "    [::textutil::adjust::adjust $Element]"
                    }
               }
               seealso {
                    puts "SEE ALSO"                    
                    puts "    [::textutil::adjust::adjust [join $value ", "]]"
               }
               example {
                    puts "EXAMPLE(S)"
                    puts ""
                    foreach Example $value {
                         foreach Line [split $Example "\n"] {
                              puts [::textutil::adjust::adjust $Line]
                         }
                    }
               }
               arguments {
                    puts "ARGUMENTS"
                    dict for {ArgKey ArgValue} $value {
                         puts "    $ArgKey"
                         puts "         [::textutil::adjust::adjust $ArgValue]"
                    }
               }
               default {
                    puts "Unknown: $value"
               }
          }
          puts ""
     }
}

proc HelpxCurrentVersion {} {
     puts 0.2
}

set HelpxDirectory [file dirname [info script]]
sqlite3 helpx-helpxdb $HelpxDirectory/helpx-helpx.db
HelpxNS::RegisterDatabase helpx-helpxdb          
